circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/relations/ecc_vm/ecc_wnaf_relation.cpp

```diff
@@ -0,0 +1,221 @@
+    /**
+     * @brief Constrain each of our scalar slice chunks (s1, ..., s8) to be 2 bits.
+     * Doing range checks this way vs permutation-based range check removes need to create sorted list + grand product
+     * polynomial. Probably cheaper even if we have to split each 4-bit WNAF slice into 2-bit chunks.
+     */
+    range_constraint_slice_to_2_bits(slices[0], std::get<0>(accumulator));
+    range_constraint_slice_to_2_bits(slices[1], std::get<1>(accumulator));
+    range_constraint_slice_to_2_bits(slices[2], std::get<2>(accumulator));
+    range_constraint_slice_to_2_bits(slices[3], std::get<3>(accumulator));
+    range_constraint_slice_to_2_bits(slices[4], std::get<4>(accumulator));
+    range_constraint_slice_to_2_bits(slices[5], std::get<5>(accumulator));
+    range_constraint_slice_to_2_bits(slices[6], std::get<6>(accumulator));
+    range_constraint_slice_to_2_bits(slices[7], std::get<7>(accumulator));
+
+    /**
+     * @brief If we are processing a new scalar (q_transition = 1), validate that the first slice is positive.
+     *        This requires us to validate slice1 is in the range [8, ... 15].
+     *        (when converted into wnaf form this maps to the range [1, 3, ..., 15]).
+     *        We do this to ensure the final scalar sum is positive.
+     *        We already know slice1 is in the range [0, ..., 15]
+     *        To check the range [8, ..., 15] we validate the most significant 2 bits (s1) are >=2
+     */
+    const auto s1_shift = View(extended_edges.precompute_s1hi_shift);
+    const auto s1_shift_msb_set = (s1_shift - 2) * (s1_shift - 3);
+    std::get<20>(accumulator) += scaled_transition * precompute_select_shift * s1_shift_msb_set;
+
+    /**
+     * @brief Convert each pair of 2-bit scalar slices into a 4-bit windowed-non-adjacent-form slice.
+     * Conversion from binary -> wnaf = 2 * binary - 15.
+     * Converts a value in [0, ..., 15] into [-15, -13, -11, -9, -7, -5, -3, -1, 1, 3, 5, 7, 9, 11 , 13, 15].
+     * We use WNAF representation to avoid case where we are conditionally adding a point in our MSM algo.
+     */
+    const auto w0 = convert_to_wnaf(slices[0], slices[1]);
+    const auto w1 = convert_to_wnaf(slices[2], slices[3]);
+    const auto w2 = convert_to_wnaf(slices[4], slices[5]);
+    const auto w3 = convert_to_wnaf(slices[6], slices[7]);
+
+    /**
+     * @brief Slice consistency check.
+     * We require that `scalar_sum` on the next row correctly accumulates the 4  WNAF slices present on the current row
+     * (i.e. 16 WNAF bits).
+     * i.e. next_scalar_sum - 2^{16} * current_scalar_sum - 2^12 * w_0 - 2^8 * w_1 - 2^4 * w_2 - w_3 = 0
+     * @note We only perform slice_consistency check when next row is processing the same scalar as the current row!
+     *       i.e. when q_transition  = 0
+     * TODO(@zac-williamson) Future work; probably don't need to convert to WNAF here. We can convert the final scalar
```

Comment:
Could you create an issue?

Reply:
https://github.com/AztecProtocol/aztec-packages/issues/2224